### ISSUE TITLE
icp fails for large files when msiExecCmd called from within pep_database_reg_data_obj_post

### ISSUE THREAD

**Issue by tedgin** [irods/irods#7183]
### iRODS Version, OS and Version

iRODS 4.2.8, CentOS 7

### What did you try to do?

I have the following rules implemented in the iRODS rule language.

```
acSetRescSchemeForCreate {
	msiSetDefaultResc("ingestRes", "null");
}

pep_database_reg_data_obj_post(*INSTANCE, *CONTEXT, *OUT, *DATA_OBJ_INFO) {
	*status = errormsg(msiExecCmd("generateuuid", "", "", "", "", *out), *msg);
	if (*status != 0) {
		writeLine('serverLog', 'generateuuid failed');
	} else {
		writeLine('serverLog', 'generateuuid succeeded');
	}
}
```

The command script `generateuuid` does nothing.

```bash
#!/usr/bin/env bash
exit 0
```

I tried to copy a 64 MiB data object using icp.

```console
icp test64MiB t
```

### Expected behavior

I expected a copy of the data object named `t` to be created.

### Observed behavior (including steps to reproduce, if applicable)

I get a `SYS_COPY_LEN_ERR`. The data object `t` was created, but it is 0 B. I verified that its physical size on disk is also 0 B.

```console
rods@dev ~$ icp test64MiB t
remote addresses: 128.196.65.197 ERROR: cpUtil: cp error for /cyverse.dev/home/rods/t, status = -27000 status = -27000 SYS_COPY_LEN_ERR
rods@dev ~$ ils -L t
  rods              0 ingestRes;ingest            0 2023-07-12.15:47 & t
        generic    /mnt/ingest/home/rods/t
``` 
```console
[irods@dev-rs-1 ~]$ ls -l /mnt/ingest/home/rods/t
-rw------- 1 irods irods 0 Jul 12 08:47 /mnt/ingest/home/rods/t
```

The following error messages were logged on the catalog provider.

```
Jul 12 08:47:51 pid:17382 NOTICE: execCmd:../../var/lib/irods/msiExecCmd_bin/generateuuid argv:
Jul 12 08:47:51 pid:17380 NOTICE: writeLine: inString = generateuuid succeeded
Jul 12 08:47:51 pid:17380 NOTICE: remoteDataCopy: rcDataCopy failed
Jul 12 08:47:51 pid:17380 NOTICE: _rsDataObjClose: size in vault 0 != target size 67108864
Jul 12 08:47:51 pid:17380 NOTICE: remoteFileClose: rcFileClose failed for 3, status = -23000
Jul 12 08:47:51 pid:17380 NOTICE: _rsDataObjClose: l3Close of 3 failed, status = -23000
Jul 12 08:47:51 pid:17380 remote addresses: 128.196.65.149, 172.16.12.135, 172.16.12.142 ERROR: [close_objects] - Failed closing data object:[-27000]
```

The following error messages were logged on the resource server.

```
Jul 12 08:47:51 pid:28994 remote addresses: 128.196.65.234 ERROR: sameHostPartialCopy: toCopy 22369621, bytesRead 0
Jul 12 08:47:51 pid:28994 remote addresses: 128.196.65.234 ERROR: sameHostPartialCopy: toCopy 22369622, bytesRead 0
Jul 12 08:47:51 pid:28994 remote addresses: 128.196.65.234 ERROR: sameHostPartialCopy: toCopy 22369621, bytesRead 0
```

I have the following resource hierarchy.

```console
rods@dev ~$ ilsresc
ingestRes:passthru
└── ingest:unixfilesystem
rods@dev ~$ ilsresc -l ingestRes
resource name: ingestRes
id: 10006
zone: cyverse.dev
type: passthru
class: cache
location: EMPTY_RESC_HOST
vault: EMPTY_RESC_PATH
free space: 
free space time: : Never
status: up
info: 
comment: 
create time: 01521060873: 2018-03-14.20:54:33
modify time: 01686168422: 2023-06-07.20:07:02
context: read=1.0;write=1.0
parent: 
parent context: 
rods@dev ~$ ilsresc -l ingest
resource name: ingest
id: 283761
zone: cyverse.dev
type: unixfilesystem
class: cache
location: 128.196.65.149
vault: /mnt/ingest
free space: 63321907200
free space time: 01689175554: 2023-07-12.15:25:54
status: up
info: 
comment: 
create time: 01554930890: 2019-04-10.21:14:50
modify time: 01689175554: 2023-07-12.15:25:54
context: minimum_free_space_for_create_in_bytes=10302260428
parent: 10006
parent context: 
```

This issues appears to be size dependent.  `icp` works fine for files with size up to 32 MiB and fails for larger files.

```console
rods@dev ~$ icp test32MiB t
rods@dev ~$ ils -L t
  rods              0 ingestRes;ingest     33554432 2023-07-12.16:01 & t
        generic    /mnt/ingest/home/rods/t
```

**Comment by trel** [comment]
do you have a 4.2.12 or 4.3.0 test instance to see if this is already fixed/healed/better?

**Comment by tedgin** [comment]
I have a single server 4.2.12 deployment.  This issue doesn't happen when I run `icp` locally on the server to copy a large data object. However, I don't have a 4.2.12 deployment with a dedicated resource server separate from the catalog service provider, so I can't prove that this is no longer a problem.

**Comment by trel** [comment]
Can you please show the file *before* the `icp`:

```
rods@dev ~$ ils -L test64MiB
```

Where is `test64MiB`?  A different resource than `ingestRes;ingest`?

What machine is the client connected to?  Presumably *not* `128.196.65.149` (the host with the `ingest` resource).

---

To reproduce, we would stand up 4.2.12 and/or 4.3.0 with two servers, one provider, one consumer...

and...

copy a large file into a not-client-connected server with that not-client-connected server having its `pep_database_reg_data_obj_post()` call `msiExecCmd()`.

**Comment by korydraughn** [comment]
@tedgin You can try using the testing environment to stand up an environment with multiple nodes.

See the stand_it_up.py script of https://github.com/irods/irods_testing_environment.

It can pull the packages for 4.2.12 or 4.3.0 from the sky for you.

**Comment by tedgin** [comment]
I can't reproduce it in 4.2.12 using the testing environment.

**Comment by korydraughn** [comment]
Were you able to stand up a zone that reflects the use-case correctly?

Did you run into any issues while trying to use the testing environment?

**Comment by tedgin** [comment]
I was able to stand up a zone that reflected the use case, but I did have a two minor issues that I had to work around.

1. When following the instructions on setting up the virtual environment under _Requirements_, `pip install docker-compose GitPython` failed because PyYAML version 5.4.1 couldn't be built due to some error in its setup.  I installed version 5.3.1, and then I was able to run the pip command.
2. The version of `docker` that I use provides a `compose` command and `docker-compose` does not exist. This mean that `docker compose` should be used to manipulate the containers instead of `docker-compose`.

**Comment by korydraughn** [comment]
I see. Please open issues in the testing environment repo for the things that blocked you.

With that said, eventually you were able to stand up the necessary zone and build confidence that the issue is not present in 4.2.12?

**Comment by trel** [comment]
@tedgin signaled success/acknowledgement as 'fixed in 4.2.12'.   Excellent.

**Comment by korydraughn** [comment]
Cool cool
